[agl] Qt-less WAM: replace qt basic data structures

nickdiego opened the pull request:
This PR is part of **Qt-less WAM** effort. Here we remove the Qt Data scructures usage from WAM (e.g: QString, QList, QMap, QFile, QDir, QVariant, etc).

### Refactoring status:
- [x] ~~QMap/QList/QPair -> STL containers~~
- [x] ~~QString -> std::string~~
- [x] ~~QStringList -> std::vector\<std::string\>~~
- [x] ~~QFile/QFileInfo/QDir -> boost::filesystem~~
- [x] ~~QUrl -> custom Url helper class (initially GLib-based)~~
- [x] ~~QVariant removal: Some unused removed + other modified to use std::string~~
- [x] ~~QLocale removal: implemented function similar to QLocale::bcp47Name()~~

### Verification status:
- [x] ~~AGL Build~~ (@nickdiego)
- [x] ~~AGL regression tests~~ (initial round done by @nickdiego)
- [ ] WebOS Build (will be handled by @jdapena when rebasing onto chromium68 branch)
- [ ] WebOS regression tests (will be handled by @jdapena when rebasing onto chromium68 branch)

#### Some remarks:
- This PR focus on common and AGL-specific code. WebOS-specific bits have been kept as much untouched as possible. So, that part might be addressed in a separate PR.
- To get this branch building correctly, a modified version of **[meta-agl-lge](uhttps://github.com/nickdiego/meta-agl-lge/tree/qtlessrl)** which adds _boost_ as WAM dependency. A PR has been sent to **meta-agl-lge** repository and its merge must be coordinated with this PR.

[[SPEC-1911] Qt-less WAM: replace qt basic data structures](https://jira.automotivelinux.org/browse/SPEC-1911)

nickdiego committed:
Replace Qt data container types by their STL counterpars

- This patch replaces Qt data structures, such as QMap, QPair, QList
by std::unordered_map, std::pair, and std::list/vector/set where applicable.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QString-less ApplicationDescription

This patch drops QString/QStringList usage from
ApplicationDescription implementation and some related
APIs.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
Add StringUtils initial implementation

StringUtils is intended to fill the gaps between
C++ standard library and Qt API's regarding string
manipulation. E.g: Qt provides a split() function
out-of-the-box while std library does not.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QString-less WebProcessManager

- This patch removes QString usage from WebProcessManager implementation.
To do that it also modifies classes that have some relationship with it,
such as WebAppManager and WebAppManagerService.
- Also, add some FIXME comments to mark points for next steps on removing
QString references from WAM.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QString-less WebAppManager

- This patch removes QString usage from WebAppManager, focusing
on removing its usage almost entirely from its public API, to do
that it also modifies classes that have some relationship with it,
such as WebApp(Base and Wayland), WebAppManagerService, WebAppFactory,
etc.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QString-less WebAppManager: refactor JS string building

This patch focus on removing QString usage in Javascript
code generation in WebApp initialization, also touching
WebPageBase and WebPageBlink implementation, modifying its
'evaluateJavascript()' method to receive an std::string instead
of QString.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QString-less WebAppManager: SystemLanguage and DeviceInfo

This patch focus on removing QString usage from SystemLanguage
and DeviceInfo API's in WebAppManager, WebAppManagerService,
DeviceInfo, WebPageBase and WebPageBlink.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QString-less WebApp/WebPage: appId

This patch focus on removing QString usage from appId APIs
in WebApp and WebPage classes. Devided to keep this as a
separate patch because appId() heavy usage all over WAM code.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QString-less WebApp: Remove leftovers

Remove QString leftovers from WebAppBase and WebAppWayland,
besides that modifies interfaces from related classes.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QVariant-less WebAppWayland

This patch refactors WebAppWayland::setWindowProperty()
method, removing its dependency on QVariant. As most of
its used was using QString and in the end the value gets
convered to a std::string the solution adopted was to modify
its parameter to a std::string.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QString-less WebPage

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QString-less NetworkStatus

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QString-less WebAppManagerConfig

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QString-less PalmSystem

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QString-less WAM: Remove leftovers

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Original Author: Abhijeet Kandalkar <abhijeet@igalia.com>
Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QUrl-less WAM: Add Url parsing code

- The main goal of this patch is to remove WAM dependency
on QUrl implementation for Url Parsing and convenience
API. To do that, we're adding our own Url API, which in this
initial version leverages GLib URI Functions [1] for basic
URI parsing and escaping.
- Also, as the code handling URLs in WAM is mostly tied to
QFile/QDir/QFileInfo manipulation, this patch also refactor
that to use Boost Filesystem module instead. It's worth noting
it depends on a modified version os meta-agl-lge (which include
boost as dependency) to build properly. This integration is still
very primitive and will be improved soon once we have WAM build
system ported from qmake to CMake.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QVariant-less PalmSystem and WebAppManagerConfig

- This patch removes QVariant usage from PalmSystem*
and WebAppManagerConfig classes.
- Completely remove PalmSystemBase::getResource() and
WebAppManagerConfig::get/setConfiguration() since they are
not being used anywhere else.
- Modify PalmSystemWebOS's pmLog* functions that were expecting
QVariants as parameter to receive std::strings instead, since they
were being used only with QStrings.
- Anyway, if we really need those methods/variables in the future
we can use boost::any to replace QVariant basic functionality.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
QLocale-less WAM: Replace QLocale by Boost.Locale

- This patch removes QLocale usage from WAM. It was being
used in WebPageBlink, to generate search path for localized
error pages.
- The previous implementation assumed QLocale::bcp47Name()
function returns an IETF Language Tag (BCP 47) string for
language string got from DeviceInfo API. Even though after
some tests with Qt API, it seems that bcp47Name() is not being
including 'script' field even when it's passed in the input string,
this path implements a boost-based genPathForLang() function that
returns a similar formatted path assuming the input string follows
the format decribed in QLocale documentation [1]

[1] http://doc.qt.io/qt-5/qlocale.html#QLocale-1

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

jdapena commented on src/util/WebAppManagerUtils.cpp:
```diff
@@ -166,3 +166,9 @@ bool WebAppManagerUtils::setGroups() {
 
+std::string WebAppManagerUtils::getEnv(const char *name)
+{
+    const char *v = getenv("TELLURIUM_NUB_PATH");
```
This is wrong, you should be checking getenv from name.
Also, when possible, we want the C++ versions of C headers. In this case, include cstdlib, and use std::getenv.

nickdiego replied:
Oops :-)
Done.

jdapena commented on src/platform/webengine/PalmSystemBlink.cpp:
```diff
@@ -14,111 +14,110 @@
+            return "true";
         else
-            return QString("false");
+            return "false";
```
return isMinimal() ? "true" : "false";

nickdiego replied:
Done.

jdapena commented on src/util/NetworkStatus.cpp:
```diff
@@ -45,20 +45,20 @@ void NetworkStatus::fromJsonObject(const Json::Value& object)
     time(&raw_time);
-    m_savedDate = QString(ctime(&raw_time));
-    m_savedDate = m_savedDate.trimmed();
+    //TODO: Trim savedDate
```
Please add trim implementation in string util, and use it here. Otherwise the logs in the error page will be harder to read.

nickdiego replied:
Done.

jdapena commented on src/core/WebAppManagerConfig.cpp:
```diff
@@ -18,6 +18,12 @@
 
+static inline std::string getEnvVar(const char *name)
+{
+    const char *v = getenv("TELLURIUM_NUB_PATH");
```
Please fix this typo

nickdiego replied:
Done.

jdapena commented on src/core/PalmSystemBase.cpp:
```diff
@@ -41,22 +45,22 @@ QVariant PalmSystemBase::getResource(QVariant a, QVariant b)
-    country = country.arg(localcountry).arg(smartServiceCountry);
+    std::stringstream jss;
+    jss << "{ \"country\": \"" << localcountry
+        << "\", \"smartServiceCountry\": \"" << smartServiceCountry << "\" }";
```
Could you take the opportunity of this change to construct JSON, then creating the string? We usually prefer to let JSON API do all the proper type checkings to construct a valid JSON content.

nickdiego replied:
This is a good point. I'll do it. But to keep the history I'll add it as a separate commit on top of the other ones. Is it ok for you?

nickdiego replied:
Done.

jdapena commented on src/core/WindowTypes.h:
```diff
@@ -1,13 +1,13 @@
+#define WT_NONE "_WEBOS_WINDOW_TYPE_NONE"
+#define WT_FLOATING "_WEBOS_WINDOW_TYPE_FLOATING"
+#define WT_UNKNOWN "_WEBOS_WINDOW_TYPE_UNKNOWN"
+#define WT_SYSTEM_UI "_WEBOS_WINDOW_TYPE_SYSTEM_UI"
```
(Not mandatory nor needed for this patchset, but...)

We usually prefer const declarations to defines like this. In this case it would be:
const char WT_CARD[]_;

And then in WindowTypes.cpp the actual strings.

nickdiego replied:
Done.

nickdiego committed:
Trim timestamp string in NetworkStatus

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

nickdiego committed:
Use const char arrays instead of macros for WindowTypes

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

jdapena commented:
Let's focus on preverification only in AGL. Once we have this ported to chromium68 branch, I can try to start proposing the patches for webOS.

andre-rosa commented on src/core/ApplicationDescription.h:
```diff
@@ -146,7 +146,7 @@ class ApplicationDescription {
 
     //Key code is changed only for facebooklogin WebApp
-    const QMap<int, QPair<int, int>>& keyFilterTable() const
+    const std::map<int, std::pair<int, int>>& keyFilterTable() const
```
Could you please take a look at the map container usage and if it doesn't need iterating on sorted order then take the opportunity to convert them to unordered_map instead? Please check the same for other occurrences.

nickdiego replied:
Ok. I'll take a look at this.
thanks for reviewing it.

nickdiego replied:
Done.

nickdiego replied:
Done.

andre-rosa commented on src/util/StringUtils.cpp:
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2008-2018 LG Electronics, Inc.
```
2018 only?

nickdiego replied:
Done.

andre-rosa commented on src/util/StringUtils.h:
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2008-2018 LG Electronics, Inc.
```
Likewise

nickdiego replied:
Done.

andre-rosa commented on src/core/WebProcessManager.cpp:
```diff
@@ -155,78 +156,72 @@ void WebProcessManager::readWebProcessPolicy()
 {
     if (!desc)
-        return QString();
+        return "";
```
return {};
or
return std::string();

Same for other places if any.

nickdiego replied:
Done.

andre-rosa commented on src/util/StringUtils.cpp:
```diff
@@ -33,3 +35,26 @@ StringList splitString(const std::string &str, char delimiter,
+                       const std::string& replaceStr)
+{
+    size_t pos = in.find(toSearch);
+    while(pos != std::string::npos) {
```
Add space between while and "("

nickdiego replied:
Done.

nickdiego committed:
Fix error handling when converting strings to numbers

- Add template function stringTo() to StringUtils.h implementing
more generic and reliable way to convert strings to number types.
Error handling follows the same approach of Qt's QString::toInt(),
using an optional argument to hold the result status for the
conversion operation.

[SPEC-1911] Qt-less WAM: replace qt basic data structures

Signed-off-by: Nick Diego Yamane <nickdiego@igalia.com>

lemketron commented:
> "This PR focus on common and AGL-specific code. WebOS-specific bits have been kept as much untouched as possible. So, that part might be addressed in a separate PR."

I saw a mention of "devmode_enabled" and TELLURIUM_NUB_PATH -- those are webOS things, so I'm curious if those mentions are in webos-only sources or if they get built for AGL as well (probably shouldn't be)?

nickdiego commented:
I've just did a first round of tests in a Raspberry Pi 3.

Initially was getting a crash when opening the first WebApp, it was a non-handled exception thrown by **std::stoi()**, most probably trying to parse an empty string. Just push a patch to uniformly fix this kind of issue, using **boost::lexical_cast()** + an helper method which handle errors similarly to Qt's **QString::toInt()** function (using an option bool pointer result parameter). After this fix, I've been able to run, for example, **Memorymatch** app smoothly, not sure if this is enough to consider merging the PR, but at least it seems to be a good starting point.
Now, I'll address @andre-rosa 's  comments.

thank you all for your reviews.

nickdiego commented:
Hi @lemketron,
> I saw a mention of "devmode_enabled" and TELLURIUM_NUB_PATH -- those are webOS things, so I'm curious if those mentions are in webos-only sources or if they get built for AGL as well (probably shouldn't be)?

I tried to focus only on Qt removal, but I remember to see and edit code using TELLURIUM_NUB_PATH env var. For example, WebAppManagerConfig implementation seems to be used by both WebOS and AGL, probably there are other classes with this kind of thing. I believe there is room for a lot of improvements, BTW we'll using [SPEC-1883](https://jira.automotivelinux.org/browse/SPEC-1883) to keep tracking of other refactors still need to be done.

Thanks for your feedback.

nickdiego commented:
Closing this in favor of #18 which targets wam's @6.agl.guppy branch and chromium68.
